File: backend/utils/logger.js
/**
 * @file        logger.js
 * @description TechTrust structured logging utility using Winston.
 *              Provides consistent logging across the entire backend.
 *              - Development: colourised console output
 *              - Production: JSON file logs (error.log + combined.log)
 *              Constitution Standard 3 — internal errors logged server-side only.
 *              Constitution Standard 4 — all significant events logged.
 * @author      Muaishaq
 * @created     2026-06-30
 * @modified    2026-06-30
 */

'use strict';

const { createLogger, format, transports } = require('winston');
const path = require('path');

// ── Log Format — Development ──────────────────────────────────────────────────
// Colourised, readable output for local development
const developmentFormat = format.combine(
  format.colorize(),
  format.timestamp({ format: 'YYYY-MM-DD HH:mm:ss' }),
  format.printf(({ timestamp, level, message, ...metadata }) => {
    let log = `${timestamp} [${level}]: ${message}`;
    if (Object.keys(metadata).length > 0) {
      log += ` ${JSON.stringify(metadata)}`;
    }
    return log;
  })
);

// ── Log Format — Production ───────────────────────────────────────────────────
// Structured JSON for log aggregation and monitoring systems
// NEVER logs sensitive data — Constitution Standard 3
const productionFormat = format.combine(
  format.timestamp(),
  format.errors({ stack: false }), // Never log stack traces in production
  format.json()
);

// ── Transports ────────────────────────────────────────────────────────────────
// Development: console only
// Production: console + rotating file logs
const buildTransports = () => {
  const transportList = [];

  if (process.env.NODE_ENV === 'production') {
    // Error log — errors only
    transportList.push(new transports.File({
      filename: path.join(__dirname, '../logs/error.log'),
      level: 'error',
      maxsize: 5242880,   // 5MB max per file
      maxFiles: 5,        // Keep last 5 rotated files
    }));

    // Combined log — all levels
    transportList.push(new transports.File({
      filename: path.join(__dirname, '../logs/combined.log'),
      maxsize: 5242880,
      maxFiles: 5,
    }));

    // Console in production too — for Railway log streaming
    transportList.push(new transports.Console({
      format: productionFormat,
    }));
  } else {
    // Development — console only with colours
    transportList.push(new transports.Console({
      format: developmentFormat,
    }));
  }

  return transportList;
};

// ── Logger Instance ───────────────────────────────────────────────────────────
const logger = createLogger({
  level: process.env.LOG_LEVEL || (process.env.NODE_ENV === 'production' ? 'warn' : 'debug'),
  format: process.env.NODE_ENV === 'production' ? productionFormat : developmentFormat,
  transports: buildTransports(),
  // Never crash the server on a logging failure
  exitOnError: false,
});

// ── Security Rule ─────────────────────────────────────────────────────────────
// These fields must NEVER appear in any log entry
// If detected, they are redacted automatically
const SENSITIVE_FIELDS = ['password', 'token', 'secret', 'authorization', 'cookie'];

/**
 * @description Removes sensitive fields from objects before logging
 * @param       {Object} data - The data object to sanitize
 * @returns     {Object} Sanitized object safe for logging
 */
const sanitizeLogData = (data) => {
  if (!data || typeof data !== 'object') return data;
  const sanitized = { ...data };
  SENSITIVE_FIELDS.forEach(field => {
    if (sanitized[field]) sanitized[field] = '[REDACTED]';
  });
  return sanitized;
};

// ── Wrapped Logger ────────────────────────────────────────────────────────────
// Sanitizes all log data before passing to Winston
// Ensures sensitive fields never reach log files
const secureLogger = {
  /**
   * @description Logs debug level messages (development only)
   * @param       {string} message - Log message
   * @param       {Object} meta - Additional metadata
   */
  debug: (message, meta = {}) => logger.debug(message, sanitizeLogData(meta)),

  /**
   * @description Logs informational messages
   * @param       {string} message - Log message
   * @param       {Object} meta - Additional metadata
   */
  info: (message, meta = {}) => logger.info(message, sanitizeLogData(meta)),

  /**
   * @description Logs warning messages
   * @param       {string} message - Log message
   * @param       {Object} meta - Additional metadata
   */
  warn: (message, meta = {}) => logger.warn(message, sanitizeLogData(meta)),

  /**
   * @description Logs error messages — never includes stack traces in production
   * @param       {string} message - Log message
   * @param       {Object} meta - Additional metadata
   */
  error: (message, meta = {}) => logger.error(message, sanitizeLogData(meta)),
};

module.exports = secureLogger;

File: ai-engine/utils/logger.py
"""
@file:        logger.py
@description: TechTrust AI Engine structured logging utility.
              Provides consistent logging across the entire AI engine.
              Sanitizes sensitive data before logging.
              Constitution Standard 3  — sensitive data never logged.
              Constitution Standard 4  — all significant events logged.
@author:      Muaishaq
@created:     2026-07-02
@modified:    2026-07-02
"""

import logging
import os
from datetime import datetime

# ── Sensitive Fields ──────────────────────────────────────────────────────────
# These fields are redacted before any log entry is written
SENSITIVE_FIELDS = [
    "api_key",
    "token",
    "password",
    "secret",
    "authorization",
    "github_token",
]

# ── Log Format ────────────────────────────────────────────────────────────────
LOG_FORMAT = "%(asctime)s [%(levelname)s] %(name)s: %(message)s"
DATE_FORMAT = "%Y-%m-%d %H:%M:%S"


def get_logger(name):
    """
    Description: Creates and returns a configured logger instance.
                 All loggers share the same format and level settings.
    Args:        name (str): Logger name — use __name__ from calling module
    Returns:     logging.Logger: Configured logger instance
    """
    logger = logging.getLogger(name)

    # Avoid adding duplicate handlers if logger already configured
    if logger.handlers:
        return logger

    # Set log level from environment
    log_level = os.getenv("LOG_LEVEL", "DEBUG")
    logger.setLevel(getattr(logging, log_level, logging.DEBUG))

    # ── Console Handler ───────────────────────────────────────────────────
    console_handler = logging.StreamHandler()
    console_handler.setFormatter(logging.Formatter(LOG_FORMAT, DATE_FORMAT))
    logger.addHandler(console_handler)

    return logger


def sanitize_log_data(data):
    """
    Description: Removes sensitive fields from data before logging.
                 Recursively sanitizes nested dictionaries.
                 Constitution Standard 3 — sensitive data never logged.
    Args:        data (dict): Data object to sanitize
    Returns:     dict: Sanitized copy of the data object
    """
    if not isinstance(data, dict):
        return data

    sanitized = {}
    for key, value in data.items():
        if key.lower() in SENSITIVE_FIELDS:
            sanitized[key] = "[REDACTED]"
        elif isinstance(value, dict):
            sanitized[key] = sanitize_log_data(value)
        elif isinstance(value, list):
            sanitized[key] = [
                sanitize_log_data(item) if isinstance(item, dict) else item
                for item in value
            ]
        else:
            sanitized[key] = value

    return sanitized


def log_scoring_event(logger, username, event, metadata=None):
    """
    Description: Logs a scoring engine event with consistent format.
                 Never logs raw GitHub data or personal identifiers.
    Args:        logger   (Logger): Logger instance from get_logger()
                 username (str):    GitHub username being scored
                 event    (str):    Event description
                 metadata (dict):   Additional context (optional)
    Returns:     None
    """
    log_entry = {
        "event": event,
        "username": username,
        "timestamp": datetime.utcnow().isoformat(),
    }

    if metadata:
        # Sanitize metadata before logging
        log_entry["metadata"] = sanitize_log_data(metadata)

    logger.info(f"SCORING EVENT | {log_entry}")